### Riddler Classic

From Emma Knight comes a unit conversion conundrum:

The sum of the factors of 36 — including 36 itself — is 91. Coincidentally, 36 inches rounded to the nearest centimeter is … 91 centimeters!

Can you find another whole number like 36, where you can “compute” the sum of its factors by converting from inches to centimeters?

### Programmatically:

- We know a number is a factor if `original mod n == 0`, so this is used with numpy arrays below. 
- I worry the value is going to be VERY large, so may need to shift to a process where I skip if a sum of factors already exceeds `original x 2.54`

In [1]:
import numpy as np 

def findFactors(n):
    """Output array of factors"""
    arr2 = np.asarray(range(1,n+1)) # build array of 1 -> n
    return np.sum(arr2[n % arr2 == 0]) # return indices where 0

assert(findFactors(36) == round(2.54 * 36,0))

In [3]:
# Confirm logic works for finding 91
n = 1
while True:
    conv_val = round(2.54 * n, 0)
    sum_fact = findFactors(n)
    
    if conv_val == sum_fact:
        print(f"Found a match: {n}")
        print(f"Conversion is: {round(2.54 * n,0)}")
        arr = np.asarray(range(1,n+1)) # build array of 1 -> n
        print(f"Factors are: {arr[n % arr == 0]}")
        print(f"Sum of factors: {sum(arr[n % arr == 0])}")
        break
    
    n += 1
    

Found a match: 36
Conversion is: 91.0
Factors are: [ 1  2  3  4  6  9 12 18 36]
Sum of factors: 91


In [10]:
# Start at 92
import time
s = time.time()
n = 92
while True:
    conv_val = round(2.54 * n, 0)
    sum_fact = findFactors(n)
    
    if conv_val == sum_fact:
        print(f"Found a match: {n}")
        print(f"Conversion is: {round(2.54 * n,0)}")
        arr = np.asarray(range(1,n+1)) # build array of 1 -> n
        print(f"Factors are: {arr[n % arr == 0]}")
        print(f"Sum of factors: {sum(arr[n % arr == 0])}")
        break
    
    n += 1

e = time.time()

print(f"Time: {e - s:.2f}")

Found a match: 378
Conversion is: 960.0
Factors are: [  1   2   3   6   7   9  14  18  21  27  42  54  63 126 189 378]
Sum of factors: 960
Time: 0.02


### Solution: 

- Factors: `[  1   2   3   6   7   9  14  18  21  27  42  54  63 126 189 378]`, sum to `960`
- Conversion: `378 * 2,54 = 960.12`, rounded = `960`

### Extra Credit:

Can we find a third? 
- My guess is these start getting pretty massive, likely smart to use something like `memoization` here

In [12]:
from functools import cache

@cache
def findFactors(n):
    """Output array of factors"""
    arr2 = np.asarray(range(1,n+1)) # build array of 1 -> n
    return np.sum(arr2[n % arr2 == 0]) # return indices where 0

# Start at 961
s = time.time()
n = 961
while True:
    conv_val = round(2.54 * n, 0)
    sum_fact = findFactors(n)
    
    if conv_val == sum_fact:
        print(f"Found a match: {n}")
        print(f"Conversion is: {round(2.54 * n,0)}")
        arr = np.asarray(range(1,n+1)) # build array of 1 -> n
        print(f"Factors are: {arr[n % arr == 0]}")
        print(f"Sum of factors: {sum(arr[n % arr == 0])}")
        break
    
    n += 1

e = time.time()

print(f"Time: {e - s:.2f}")

Found a match: 49600
Conversion is: 125984.0
Factors are: [    1     2     4     5     8    10    16    20    25    31    32    40
    50    62    64    80   100   124   155   160   200   248   310   320
   400   496   620   775   800   992  1240  1550  1600  1984  2480  3100
  4960  6200  9920 12400 24800 49600]
Sum of factors: 125984
Time: 54.19


### Can We Make Our Function Faster? 

Do we typically see that we exceed our initial value with factors or that we can't hit the amount? 